File: apps/dashboard/src/lib/secrets.ts
// Secret hardening (claude.md §6 / prompt §7): no working default secrets.
// Reject known placeholders and enforce a minimum-entropy / length floor. The
// dashboard side mirrors the Python startup check for the ingest token.

import crypto from "node:crypto";

// Known placeholder fragments that must never be accepted as a real secret.
const PLACEHOLDERS = [
  "changeme",
  "change-me",
  "bitte-ersetzen",
  "replace",
  "your-key-here",
  "your-token",
  "example",
  "placeholder",
  "todo",
  "xxx",
  "secret",
  "dummy",
];

const MIN_TOKEN_LENGTH = 24;

export interface SecretCheck {
  ok: boolean;
  reason?: string;
}

/** Validate a shared-secret style token: present, long enough, not a placeholder. */
export function checkToken(value: string | undefined): SecretCheck {
  if (value === undefined || value.trim() === "") {
    return { ok: false, reason: "missing" };
  }
  const v = value.trim();
  if (v.length < MIN_TOKEN_LENGTH) {
    return { ok: false, reason: `too short (< ${MIN_TOKEN_LENGTH} chars)` };
  }
  const lower = v.toLowerCase();
  for (const p of PLACEHOLDERS) {
    if (lower.includes(p)) return { ok: false, reason: `looks like a placeholder ("${p}")` };
  }
  // Crude entropy floor: require a mix of character classes.
  const classes = [/[a-z]/, /[A-Z0-9]/, /[-_]/].filter((re) => re.test(v)).length;
  if (classes < 2) return { ok: false, reason: "insufficient character variety" };
  return { ok: true };
}

/** The configured ingest token, or null when it fails the hardening check. */
export function getIngestToken(): string | null {
  const raw = process.env.DIE_FIRMA_INGEST_TOKEN;
  return checkToken(raw).ok ? (raw as string).trim() : null;
}

/** Constant-time comparison to avoid leaking the token via timing. */
export function tokenMatches(provided: string | null, expected: string): boolean {
  if (provided === null) return false;
  const a = Buffer.from(provided);
  const b = Buffer.from(expected);
  if (a.length !== b.length) return false;
  return crypto.timingSafeEqual(a, b);
}

/** Pull a bearer-style token from request headers. */
export function tokenFromRequest(req: Request): string | null {
  const direct = req.headers.get("x-die-firma-token");
  if (direct !== null && direct.trim() !== "") return direct.trim();
  const auth = req.headers.get("authorization");
  if (auth !== null && auth.toLowerCase().startsWith("bearer ")) {
    return auth.slice(7).trim();
  }
  return null;
}

File: services/orchestrator/die_firma/secrets.py
"""Secret hardening for the Python side (claude.md §6 / prompt §7).

Mirrors the dashboard's check: reject known placeholders, enforce a minimum
length / character-variety floor, validate the Anthropic key format. Startup
aborts when a required secret fails. 100% coverage target.
"""

from __future__ import annotations

PLACEHOLDERS = (
    "changeme",
    "change-me",
    "bitte-ersetzen",
    "replace",
    "your-key-here",
    "your-token",
    "example",
    "placeholder",
    "todo",
    "xxx",
    "secret",
    "dummy",
)

MIN_TOKEN_LENGTH = 24


class SecretError(ValueError):
    """A required secret is missing or looks like a placeholder."""


def _contains_placeholder(value: str) -> str | None:
    lower = value.lower()
    for p in PLACEHOLDERS:
        if p in lower:
            return p
    return None


def check_token(value: str | None, *, name: str = "token") -> str:
    """Validate a shared-secret style token; return it trimmed or raise."""
    if value is None or value.strip() == "":
        raise SecretError(f"{name} is missing")
    v = value.strip()
    if len(v) < MIN_TOKEN_LENGTH:
        raise SecretError(f"{name} is too short (< {MIN_TOKEN_LENGTH} chars)")
    hit = _contains_placeholder(v)
    if hit is not None:
        raise SecretError(f"{name} looks like a placeholder ({hit!r})")
    classes = sum(
        1
        for ok in (
            any(c.islower() for c in v),
            any(c.isupper() or c.isdigit() for c in v),
            any(c in "-_" for c in v),
        )
        if ok
    )
    if classes < 2:
        raise SecretError(f"{name} has insufficient character variety")
    return v


def check_anthropic_key(value: str | None) -> str:
    """Validate an Anthropic API key (format + not a placeholder)."""
    if value is None or value.strip() == "":
        raise SecretError("ANTHROPIC_API_KEY is missing")
    v = value.strip()
    if not v.startswith("sk-ant-"):
        raise SecretError("ANTHROPIC_API_KEY must start with 'sk-ant-'")
    hit = _contains_placeholder(v)
    if hit is not None:
        raise SecretError(f"ANTHROPIC_API_KEY looks like a placeholder ({hit!r})")
    if len(v) < MIN_TOKEN_LENGTH:
        raise SecretError("ANTHROPIC_API_KEY is too short")
    return v
